lib/TPP/GPU/LinalgToXeGPU.cpp

```diff
@@ -0,0 +1,1402 @@
+                                          PatternRewriter &rewriter,
+                                          unsigned maxDims = 2) {
+  auto type = dyn_cast<MemRefType>(operand.getType());
+  if (!type) {
+    return rewriter.notifyMatchFailure(
+        linalgOp, "Expect memref operand for XeGPU lowering");
+  }
+
+  if (type.getShape().size() > maxDims) {
+    return rewriter.notifyMatchFailure(
+        linalgOp, "Too high dimensionality for XeGPU operations");
+  }
+
+  auto strides = utils::getStaticStrides(operand);
+
+  if (failed(strides)) {
+    return rewriter.notifyMatchFailure(
+        linalgOp, "Expect static strides for XeGPU lowering");
+  }
+  if (strides->back() != 1) {
+    return rewriter.notifyMatchFailure(linalgOp,
+                                       "Expect unit stride in the innermost "
+                                       "dimension for XeGPU operations");
+  }
+
+  return success();
+}
+
+// Match and, if possible, lower a generic operation to an XeGPU compatible op.
+// Returns the result of the lowered op or nullopt, otherwise.
+static std::optional<Value> lowerGenericOp(linalg::GenericOp genericOp,
+                                           ArrayRef<Value> operands,
+                                           VectorType resType,
+                                           PatternRewriter &rewriter) {
+  Location loc = genericOp.getLoc();
+
+  // Expect operands to be already loaded vectors.
+  for (auto operand : operands) {
+    if (!isa<VectorType>(operand.getType()))
+      return std::nullopt;
+  }
+
+  if (structured_match::utils::isTwoDReluOp(genericOp, /*operands=*/nullptr)) {
+    assert(operands.size() == 1 &&
+           "Invalid number of operands for generic 2D ReLU");
+
+    auto eltType = resType.getElementType();
+    Value zeroConst;
+
+    if (isa<FloatType>(eltType)) {
+      auto floatType = cast<FloatType>(eltType);
+      zeroConst = rewriter.create<arith::ConstantFloatOp>(
+          loc, APFloat::getZero(floatType.getFloatSemantics()), floatType);
+    } else if (isa<IntegerType>(eltType)) {
+      zeroConst = rewriter.create<arith::ConstantIntOp>(loc, 0, eltType);
+    } else {
+      // Unhandled type. Bail out.
+      return std::nullopt;
+    }
+
+    auto zeroVec =
+        rewriter.create<vector::BroadcastOp>(loc, resType, zeroConst);
+
+    return rewriter
+        .create<arith::MaximumFOp>(loc, resType, operands[0], zeroVec)
+        .getResult();
+  }
+
+  if (structured_match::utils::isTwoDAddOp(genericOp, /*operands=*/nullptr)) {
+    assert(operands.size() == 2 &&
+           "Invalid number of operands for generic 2D add");
+    return rewriter
+        .create<arith::AddFOp>(loc, resType, operands[0], operands[1])
+        .getResult();
+  }
+
+  return std::nullopt;
+}
+
+// Lower an elementwise operation to an XeGPU compatible op.
+// Returns the result of the lowered op or nullopt, otherwise.
+static std::optional<Value> lowerEltwiseOp(linalg::LinalgOp linalgOp,
```

Comment:
Can you not just generalize (upstream) and then only use the above method?

Reply:
I think something comparable (Linalg -> arith) can be achieved using Linalg vectorizer and vector utilities. So, such explicit conversion like here shouldn't be needed. If I notice something missing there, then it can be fixed directly upstream.